### ISSUE TITLE
関連Issueの作成

### ISSUE THREAD

**Issue by susumutomita** [susumutomita/diy-gpt#2]
<html><head></head><body><p>以下は <strong>Python + NumPy/CuPy</strong> で “GPT-2 ミニ” をスクラッチ実装するための最終版ロードマップと GitHub Issue 分割例です。前回 Ruby 向けに書いた内容を <strong>Python 前提</strong> に最適化し、余分な説明は省いてあります。</p>
<hr>
<h2>プロジェクト方針（Python 版）</h2>

項目 | 方針
-- | --
数値演算 | NumPy を基本にし、GPU を使える環境なら cupy もオプションとしてサポート（同 API）。
自動微分 | “理解” が目的なので 手書きの誤差逆伝播 を実装。最終段階で jax/tinygrad などへ差し替えて挙動比較しても良い。
依存関係 | numpy >=1.26, cupy (任意), pytest, black, isort, ruff
構成 | src/ 配下にモジュールを分割。tests/ に pytest ケース。
最小モデル | 4 layers / 4 heads / d_model = 256（約 5 M params）から開始。
実験データ | “Tiny Shakespeare” 1 MB（文字レベル → BPE に発展）。

<hr>
<h2>GitHub Issue テンプレート</h2>
<blockquote>
<p><strong>使い方</strong> : リポジトリ作成後、下記をコピペして Issue を立てるだけ。チェックリストは適宜追加／削除してください。</p>
</blockquote>
<hr>
<h3>Issue 0 : Python 環境セットアップ</h3>
<pre><code class="language-md">- [ ] poetry / pip-tools で lock ファイルを生成
- [ ] pre-commit で black・ruff・pytest を自動実行
- [ ] GitHub Actions で `pytest -q` を CI
</code></pre>
<h3>Issue 1 : Byte-Pair Encoding トークナイザ</h3>
<pre><code class="language-md">- [ ] `scripts/train_bpe.py` で語彙・マージ規則を学習
- [ ] `src/tokenizer.py` に `encode`, `decode`
- [ ] pytest: encode→decode が恒等写像になること
</code></pre>
<h3>Issue 2 : 埋め込み + 位置エンコーディング層</h3>
<pre><code class="language-md">- [ ] `TokenEmbedding(vocab_size, d_model)`
- [ ] `PositionalEmbedding(max_len, d_model)`  (学習可能)
- [ ] forward: token_emb + pos_emb
- [ ] pytest: 出力 shape / 同一入力の再現性
</code></pre>
<h3>Issue 3 : Scaled Dot-Product Attention（マスク付き）</h3>
<pre><code class="language-md">- [ ] `attention(Q, K, V, mask=None)` を NumPy で実装
- [ ] causal mask (`np.triu(...)`) 対応
- [ ] 数値微分で勾配テスト
</code></pre>
<h3>Issue 4 : Multi-Head Attention モジュール</h3>
<pre><code class="language-md">- [ ] `MultiHeadAttention(d_model, n_heads)` クラス
- [ ] 重み初期化: Xavier uniform
- [ ] pytest: n_heads=1 が Issue 3 と一致
</code></pre>
<h3>Issue 5 : Feed-Forward Network (GELU)</h3>
<pre><code class="language-md">- [ ] `FeedForward(d_model, d_ff=4*d_model)`
- [ ] GELU 近似 `0.5*x*(1+tanh(√(2/π)*(x+0.044715*x³)))`
- [ ] pytest: shape / 勾配テスト
</code></pre>
<h3>Issue 6 : LayerNorm</h3>
<pre><code class="language-md">- [ ] ε=1e-5, learnable γ, β
- [ ] forward/backward を NumPy で
- [ ] pytest: 出力平均≈0, 分散≈1
</code></pre>
<h3>Issue 7 : Transformer Block 組み立て</h3>
<pre><code class="language-md">- [ ] Pre-LN 構造 (LN→MHA→残差→LN→FFN→残差)
- [ ] Dropout はダミー (後で実装)
- [ ] pytest: アイデンティティ初期化で入出力一致
</code></pre>
<h3>Issue 8 : GPT-Mini モデル全体</h3>
<pre><code class="language-md">- [ ] `GPTMini(n_layers=4, ...)`
- [ ] トークナイザ→Embedding→[Block×n]→LN→Linear(vocab)
- [ ] softmax は損失計算側で
- [ ] pytest: forward shape (batch, seq, vocab)
</code></pre>
<h3>Issue 9 : 損失関数 &amp; オプティマイザ</h3>
<pre><code class="language-md">- [ ] `cross_entropy(logits, target)`
- [ ] Adam 手書き実装 (β₁=0.9, β₂=0.999, ε=1e-8)
- [ ] pytest: 数値微分で勾配検証
</code></pre>
<h3>Issue 10 : 訓練ループ &amp; Tiny Shakespeare 実験</h3>
<pre><code class="language-md">- [ ] `train.py` (CLI: --epochs, --device cpu|gpu)
- [ ] ランダムシード/チェックポイント保存
- [ ] 学習曲線を Matplotlib に保存
- [ ] 生成サンプルを `samples/` に出力
</code></pre>
<h3>Issue 11 : モデルサイズ拡張 &amp; CuPy / JAX ベンチ</h3>
<pre><code class="language-md">- [ ] d_model = 512 / 8 heads / 6 layers
- [ ] CuPy backend に切替 benchmark
- [ ] オプションで JAX 実装と速度比較
</code></pre>
<h3>Issue 12 : ドキュメント &amp; 理論メモ</h3>
<pre><code class="language-md">- [ ] `docs/architecture.md` に数式と実装対応表
- [ ] `docs/backprop.md` に手書き勾配導出
- [ ] API ドキュメント自動生成 (pdoc etc.)
</code></pre>
<h3>Issue 13 (Stretch) : 推論最適化</h3>
<pre><code class="language-md">- [ ] 重みを fp16 で保存 &amp; 量子化 (int8) 実験
- [ ] キャッシュ付き高速デコード (KV-cache)
</code></pre>
<hr>
<h2>進め方ヒント</h2>
<ol>
<li>
<p><strong>前向き計算が全て通る</strong> ことを最優先。</p>
</li>
<li>
<p>手書き勾配は最初はバグりやすいので、勾配テスト（数値微分）を必ず走らせる。</p>
</li>
<li>
<p>CPU 学習が遅ければ CuPy → GPU、あるいは JAX/XLA へ移行。</p>
</li>
<li>
<p>小規模コーパスで過学習 → 出力が “丸暗記” できれば実装正しい証拠。</p>
</li>
<li>
<p>すべてのモジュールで <strong>shape assert</strong> を書くとデバッグが楽。</p>
</li>
</ol>
<p>これで Python 版のスクラッチ実装をスムーズに進められるはずです。楽しんでください！</p></body></html>
